store/streaming/file/service.go

```diff
@@ -20,262 +22,140 @@ var _ baseapp.StreamingService = &StreamingService{}
-			fss.stateCache = nil
-			fss.stateCacheLock.Unlock()
-			return err
-		}
-	}
-	// reset cache
-	fss.stateCache = nil
-	fss.stateCacheLock.Unlock()
-	// write res to file
-	lengthPrefixedResBytes, err := fss.codec.MarshalLengthPrefixed(&res)
-	if err != nil {
-		return err
-	}
-	_, err = dstFile.Write(lengthPrefixedResBytes)
-	return err
+// ListenEndBlock satisfies the baseapp.ABCIListener interface
+// It writes the received EndBlock request and response and the resulting state changes
+// out to a file as described in the above the naming schema
+func (fss *StreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) (rerr error) {
+	fss.blockMetadata.RequestEndBlock = &req
+	fss.blockMetadata.ResponseEndBlock = &res
+	return nil
 }
 
-func (fss *StreamingService) openDeliverTxFile() (*os.File, error) {
-	fileName := fmt.Sprintf("block-%d-tx-%d", fss.currentBlockNumber, fss.currentTxIndex)
+func (fss *StreamingService) ListenCommit(ctx sdk.Context, res abci.ResponseCommit) error {
+	fss.blockMetadata.ResponseCommit = &res
+
+	// write to target file atomically
+	fileName := fmt.Sprintf("block-%d", fss.currentBlockNumber)
 	if fss.filePrefix != "" {
 		fileName = fmt.Sprintf("%s-%s", fss.filePrefix, fileName)
 	}
-	fss.currentTxIndex++
-	return os.OpenFile(filepath.Join(fss.writeDir, fileName), os.O_CREATE|os.O_WRONLY, 0o600)
+	targetFile := path.Join(fss.writeDir, fileName)
+	tmpFile := path.Join(fss.writeDir, "tmpfile")
+	if err := fss.writeToFile(tmpFile); err != nil {
+		return err
+	}
+	return atomic.ReplaceFile(tmpFile, targetFile)
 }
 
-// ListenEndBlock satisfies the baseapp.ABCIListener interface
-// It writes the received EndBlock request and response and the resulting state changes
-// out to a file as described in the above the naming schema
-func (fss *StreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) (rerr error) {
-	// generate the new file
-	dstFile, err := fss.openEndBlockFile()
+func (fss *StreamingService) writeToFile(path string) (rerr error) {
+	fp, err := os.OpenFile(filepath.Clean(path), os.O_CREATE|os.O_WRONLY, 0o600)
 	if err != nil {
 		return err
 	}
+
 	defer func() {
-		cerr := dstFile.Close()
+		err := fp.Close()
 		if rerr == nil {
-			rerr = cerr
+			rerr = err
 		}
 	}()
 
-	// write req to file
-	lengthPrefixedReqBytes, err := fss.codec.MarshalLengthPrefixed(&req)
+	bz, err := fss.codec.MarshalLengthPrefixed(&fss.blockMetadata)
 	if err != nil {
 		return err
 	}
-	if _, err = dstFile.Write(lengthPrefixedReqBytes); err != nil {
+	if _, err = fp.Write(bz); err != nil {
 		return err
 	}
-	// write all state changes cached for this stage to file
-	fss.stateCacheLock.Lock()
-	for _, stateChange := range fss.stateCache {
-		if _, err = dstFile.Write(stateChange); err != nil {
-			fss.stateCache = nil
-			fss.stateCacheLock.Unlock()
-			return err
+
+	for _, listener := range fss.listeners {
```

Comment:
I can't imagine groking these resulting streamed files per block is ideal.

1. You're writing a single encoded object of type `blockMetadata`.
2. Them you're writing (multiple) `entry` objects, which are of type KVStore.

How does an operator or client use these files effectively?

I think a better UX would be buffer all the writes into a single type and write that single type. E.g. adding `[]entry` to the existing metadata and also, not calling it metadata.

This might be to drastic of a change, but I do think it's a better UX.

Reply:
[here](https://github.com/crypto-org-chain/cronos/blob/b78d8eb146219e3558505dbe574c8fbe8d4247c6/integration_tests/test_streamer.py#L16) is how I decode the file in my integration test.
All the items are length delimited, it starts with a `BlockMetadata`, the rest are `StoreKVPairs`s.

Reply:
[here](https://github.com/crypto-org-chain/cronos/blob/134980ddc921c58d4cfe3dfbf470185ff70db730/versiondb/sync.go#L17) is parse the `StoreKVPair` in golang, skip the metadata part.

Reply:
Yeah, I'm sure it works. I think it's just ugly personally.

Reply:
I think length delimited is better for streamlined decoding, an alternative is save metadata and kv writes in two files, that sounds good too.

Reply:
I'll write two files for metadata and data, the data file is a stream of length delimited KVStore, the meta file is a single object `BlockMetadata`.